### ISSUE TITLE
Improved pattern for mixin and trait

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/phet-info#143]
From developer meeting discussion (https://github.com/phetsims/phet-core/issues/54) @samreid found a potentially better way for us to implement mixin/ trait pattern. Here is the resource

https://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/

@jonathanolson said he would be interested in evaluating.

**Comment by jonathanolson** [comment]
This looks quite promising, and "feels" cleaner, with the exception of constructors (we'd need to use rest/spread operators for constructors instead of initializers, which is fine) and the fact that we'd presumably need to either (a) turn almost everything into ES5 classes, for the Node mixins, or (b) figure out something that works with both.

Follow-up article is needed: https://justinfagnani.com/2016/01/07/enhancing-mixins-with-decorator-functions/, and the helper library is at https://github.com/justinfagnani/mixwith.js.

**Comment by samreid** [comment]
I agree this looks like the appropriate way to proceed.  The class-factory pattern is 

```js
const MyMixin = Superclass => class extends Superclass {
  // mixin methods go here
};

class A extends MyMixin(MyBaseClass) {}
```

Our existing mixins are:
* Poolable
* AccessibleSlider
* PaintableStatelessDrawable
* PaintableStatefulDrawable
* CircleStatefulDrawable
* ImageStatefulDrawable
* PathStatefulDrawable
* RectangleStatefulDrawable
* TextStatefulDrawable
* Leaf
* Paintable
* EnabledNode
* EnabledPhetioObject
* AccessibleNumberSpinner
* AccessibleValueHandler

So it seems to move toward the class-factory pattern, we would need to change every type that these are mixed into to be `class` definitions.  Should we chip away at this?

Also, we currently have `AccessibleSlider` "extends" `AccessibleValueHandler`--how would that be done with the class-factory pattern?

Tagging for discussion at developer meeting.

**Comment by jonathanolson** [comment]
Also, we can create inherit-based versions of this for use with our current inherit-based types.

**Comment by mattpen** [comment]
Dev Meeting: This seems promising, but @ariel-phet said it should be on hold until density is published.

**Comment by jonathanolson** [comment]
Somewhat ironically, I almost needed to use this style of mixin to properly handle a case in Density (so it came first). Tagging to view/discuss during dev meeting.

**Comment by samreid** [comment]
Today @jonathanolson described how he is using this pattern in Density & Buoyancy in `DensityAndBuoyancyModel.js`, and reported that it is working well.  He said it is a likely candidate for replacing old mixin patterns, but may take more effort to use with `inherit` instead of `class`.

@jbphet said this may be an opportunity for us to upgrade some types to `class` instead of `inherit`.  Perhaps `Vector2`.

The current proposal is to use this pattern to create new mixins.  Then when we have more experience we can decide whether to backport this pattern to existing mixin patterns.  @jonathanolson invited developers to reach out to him for collaboration when starting new mixins.

**Comment by ariel-phet** [comment]
Calendar reminder made

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/sun/issues/585, I've been tasked with converting components to use EnabledComponent.  It (and its sub-mixins) are apparently "old style".   Do we really want to convert a bunch of things to use an old pattern?  Changing the status of this issue and labeling for developer meeting with high priority.

**Comment by samreid** [comment]
The improved pattern for mixin requires `class`.  Ideally we can migrate everything soon, including Node and PhetioObject as we proceed in https://github.com/phetsims/tasks/issues/1050.  Using the old pattern for mixin is acceptable until then.

Not sure if this is the right issue for the next idea: Now that Node supports TinyProperty, and has established support for passing in properties, PhET-iO support (linked elements & instrumentation), perhaps we should move `enabledProperty` directly to `Node`.

**Comment by mattpen** [comment]
@ariel-phet will tag the sim conversion issue here after it is created and put this issue on hold.

**Comment by ariel-phet** [comment]
The sim conversion issue is https://github.com/phetsims/phet-info/issues/143

**Comment by jbphet** [comment]
The conversion of all sims to ES6 is now complete, so I'm removing the "on-hold" label and adding the developer meeting label.

**Comment by ariel-phet** [comment]
MK: If we do this work we can get rid of `extend` usage

JO: Doing the work in this issue is a good idea but keeping `extend` around might be useful. For instance, if you need to cases where class inheritance does not apply.

**Comment by ariel-phet** [comment]
@jonathanolson will create a chip-away issue (with a checklist in the style of our other chip-away issues) to enumerate where the old pattern is used.

**Comment by samreid** [comment]
Developer Meeting:

@zepumph volunteered to help identify existing cases that can be ported to the new style, thanks!

**Comment by zepumph** [comment]
https://github.com/phetsims/phet-info/issues/154 has been created. @jonathanolson what else would you like to see for this issue? Do we need to update documentation or processes at all?

**Comment by zepumph** [comment]
Actually, @jonathanolson, I have one question about how to convert traits. The main pattern we use when adding features to Node (and often Node subtypes) is to mixin the features in that parent class before exporting the class, for example with ParallelDOM in Node

https://github.com/phetsims/scenery/blob/e4a96cad0bec6eed5cfd5b9d490257de4eed92c0/js/nodes/Node.js#L6704

How would we compose this with the new pattern? Would it be something like this?

```diff
Index: js/nodes/Node.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/nodes/Node.js b/js/nodes/Node.js
--- a/js/nodes/Node.js	(revision e4a96cad0bec6eed5cfd5b9d490257de4eed92c0)
+++ b/js/nodes/Node.js	(date 1613074967208)
@@ -6701,7 +6701,7 @@
 scenery.register( 'Node', Node );
 
 // Node is composed with this feature of Interactive Description
-ParallelDOM.compose( Node );
+Node = ParallelDOM( Node );
 
 // @public {IOType}
 Node.NodeIO = new IOType( 'NodeIO', {
```

**Comment by jonathanolson** [comment]
The new pattern is like https://github.com/phetsims/scenery/blob/master/js/display/drawables/PaintableStatelessDrawable.js. It would probably be beneficial to work on this together.

It's memoized (so if it's mixed into the super of the same type multiple times, it won't re-do things), and notably that function returns a class. Super and dispose can be overridden just like a class (in this case, it's poolable, so initialize was overridden).

Should we work through that together?

**Comment by zepumph** [comment]
Thanks for pointing out that example. I think pairing would be nice, as I still have the following questions about how to convert ParallelDOM

* Doing the pattern like `class Node extends ParallelDOM( PhetioObject ) { ... }` feels weird and wrong to me since ParallelDOM is a trait only to be used with Node and its subtypes (yet it gets mixed in sorta as a "before" step to gaining Node functionality. For example, what if ParallelDOM had the need to overwrite a method of Node? 
* A similar question is with how we would add mutatorKeys. Right now, it doesn't seem like Node is prepared to take the supertype's mutatorKeys into account, so might the PDOM keys just get overwritten?

**Comment by pixelzoom** [comment]
Over in #154, we decided to revisit this issue. @samreid mentioned lack of IDE support for mixins (new and old patterns) and suggested that we investigate using compostion instead of mixins. 

Here's an article that enumerates pitfalls of mixin, and discusses moving from mixins to "automated composition":
https://raganwald.com/2016/07/20/prefer-composition-to-inheritance.html

**Comment by zepumph** [comment]
@jessegreenberg's comment from over in https://github.com/phetsims/phet-info/issues/154#issuecomment-790814051

> Discussed during 3/4/21 meeting -
> 
> JO: There is an issue with the new pattern for mixin/trait where there isn't IDE support.
> SR: Maybe we should transtion away from mixins?
> MK: If IDE support is the only bad part of this, can we make our own pluggin to support it?
> 
> Sounds like there are loose ends still and we aren't ready to move forward with the new pattern yet. We will pick it up next time.


JO also mentioned a new pattern called [Entity component system](https://en.wikipedia.org/wiki/Entity_component_system)

**Comment by zepumph** [comment]
We then found that with the new mixin pattern, WebStorm was smart enough in `Image` (which `extends Imageable( Node )`), to auto complete Node methods and api.

**Comment by jessegreenberg** [comment]
3/4/21 - 

@samreid is going to investigate the new mixin/trait pattern to see if we can find IDE or other issues before converting. Assigning to him, and we will discuss again after.

**Comment by samreid** [comment]
To investigate, I compared `class FlowBox extends WidthSizable( HeightSizable( Node ) ) {` with `Node`.  

Testing auto-complete on a Node instance shows correct completion options:

![image](https://user-images.githubusercontent.com/679486/110354234-2fc3ac80-7ff5-11eb-9a50-a9dccc10dfbf.png)

Testing auto-complete on a FlowBox doesn't know what to do:

![image](https://user-images.githubusercontent.com/679486/110354309-410cb900-7ff5-11eb-9366-c551c0b2ad33.png)

Node navigation takes you to the declaration of `getSafeTransformedVisibleBounds`.  If you fill in the method by hand, then flowBox will take you there as well, because that method name is unique across our project.  However, if you create another `getSafeTransformedVisibleBounds`, then Node will navigate correctly, but flowBox will show options:

![image](https://user-images.githubusercontent.com/679486/110356063-1c194580-7ff7-11eb-9904-4db2187cb905.png)

To summarize, the mixin pattern doesn't autocomplete to mixed or parent methods, but if you type in the code yourself, it will highlight and navigate according to untyped "loose" Javascript patterns.  Is this a good enough reason to avoid this pattern?  In the absence of a superior alternative mixin pattern, I'd say not.  But it could encourage us to use inheritance or composition rather than mixins where possible.  For instance, `Paintable` is currently only mixed into `Path` and `Text` which means we could create `PaintableNode` extends `Node` (unless we plan for future Paintable things to be non-Nodes).

**Comment by pixelzoom** [comment]
FYI, I had an excellent experience converting a mixin to composition in https://github.com/phetsims/ph-scale/issues/210.  This was the only place that I was using mixin in code that I'm responsible for.  It always felt unnecessarily complicated, somewhat obfuscated.  I'm very happy with the switch to composition.

**Comment by samreid** [comment]
I think we would maintain the API via forwarding methods, like:

```
get fill(){ return this.paintable.fill;}
```

I'm not opposed to duplication of these adapter-style methods because they are simple, clear, and unlikely to hide bugs.

**Comment by jonathanolson** [comment]
> I'm not opposed to duplication of these adapter-style methods because they are simple, clear, and unlikely to hide bugs.

But any addition requires finding everywhere it's used and applying refactoring. It also adds runtime overhead, additional boilerplate code, etc.

**Comment by samreid** [comment]
Summary for April dev meeting:

1. @samreid confirmed that the proposed mixin pattern `class FlowBox extends WidthSizable( HeightSizable( Node ) )` does not support autocomplete/highlighting/navigation.
2. @pixelzoom affirmed that composition is preferable to mixins when it is possible.  @jonathanolson points out that we still want to maintain the same API even if we use composition, and I agree.

Proposed remaining work for this issue:
1. Convert anything using the old `mixInto` pattern to the new pattern?  Poolable, AccessibleSlider, Leaf, AccessibleNumberSpinner, AccessibleValueHandler, ValueGestureControl.  Who will convert these?  What steps are necessary to convert?
2. Paintable is currently only mixed into Path and Text which means we could create PaintableNode extends Node?

**Comment by zepumph** [comment]
Today we noted that the new mixin pattern is a regression on Webstorm syntax highlighting for the supertypes. The main blocking issue is here is that Webstorm syntax highlighting.

JB and CK would like to have syntax highlighting, and don't use mixins very often, but don't feel like highlighting is  a deal breaker to the new feature. 

CM: Messing up the supertype's api is a big problem to me. 

SR: prefers patterns that support our primary IDE's workflow (syntax highlighting/autocomplete/etc)

JO: wants to use this new mixin pattern because besides webstorm's support it is preferable in every way, and factor's out a lot of code.

CM: If we need mixins so much, we are using the wrong language.

Possible ways forward:
* Find a way for our mixin pattern to support Webstorm
* Duplicate boilerplate in every spot we use the mixin (to support API forwarding to composition)
* Brainstorm other mixin patterns that would work better for Webstorm.
* See if `@mixin` jsdoc helps in supporting this.

**Comment by zepumph** [comment]
@samreid will look into using jsdoc, and @jonathanolson will investigate the others.

**Comment by samreid** [comment]
I investigated `@mixin` and `@mixes` in an isolated context, using this code example:

```js
// Copyright 2014-2020, University of Colorado Boulder

/**
 * @mixes Administrator
 */
class Person {}

/**
 * @mixin
 */
class Administrator {
  goAdministrate() {}
}

/**
 * @mixin
 */
class Fisher {
  goFish() {}

  goDoSomething() {}
}

/**
 * @mixin
 */
class UnrelatedClass {
  goFish() {}
}

/**
 * @mixin
 */
class OtherType {
  goFish() {}

  goHide() {}
}

const p = new Person();
console.log( p );

const c = new UnrelatedClass();

p.goFish();
```

For autocomplete, the `@mixes` annotation doesn't seem to do anything.  It wants to autocomplete anything in the file.

![image](https://user-images.githubusercontent.com/679486/115089399-6089e200-9ecf-11eb-90a3-0ac684a71925.png)

However, for navigation, it's a different story.  If I put `Person @mixes Fisher`, navigation navigates `goFish` to the `Fisher` implementation.  Likewise, if I put `Person @mixes OtherType`, then navigation navigates `goFish` to the `OtherType` implementation.  Next I will try these experiments on our actual types.

**Comment by samreid** [comment]
I tried instantiating a new FlowBox at the bottom of FlowBox.js, and added `@mixes WidthSizable` like so:

```js
/**
 * @mixes WidthSizable
 */
class FlowBox extends WidthSizable( HeightSizable( Node ) ) {
```

```js
const f = new FlowBox();
```

It did not have any ideas about autocompleting `preferredWidthProperty`:

![image](https://user-images.githubusercontent.com/679486/115089892-94193c00-9ed0-11eb-8579-3173c16a0147.png)


If I removed the `memoize` at the WidthSizable declaration, then it does better:

![image](https://user-images.githubusercontent.com/679486/115089912-a6937580-9ed0-11eb-93a5-44a420dd930b.png)

However, removing the `@mixes` did not disrupt this in any way.

In short, these JSDoc issues may help slightly.

**Comment by samreid** [comment]
I tried 

```js
/**
 * @mixes Poolable
 */
class ImageSVGDrawable extends ImageStatefulDrawable( SVGSelfDrawable ) {
```

and it was unable to autocomplete freeToPool:

![image](https://user-images.githubusercontent.com/679486/115090201-7bf5ec80-9ed1-11eb-9642-b0ba1e10fbc1.png)

**Comment by jonathanolson** [comment]
> If I removed the memoize at the WidthSizable declaration, then it does better

So... potentially we could get rid of the memoize? It's designed to be better for memory/performance, but I haven't measured the effects.

**Comment by samreid** [comment]
> CM: If we need mixins so much, we are using the wrong language.

An initial look at Typescript seems very promising, see https://github.com/phetsims/tasks/issues/987.  Perhaps we don't need to spend too much time finding the perfect long-term solution for this issue if we can migrate to typescript instead.

**Comment by samreid** [comment]
As reported in https://github.com/phetsims/chipper/issues/1049#issue-920727880,  I confirmed the pattern described in https://www.typescriptlang.org/docs/handbook/mixins.html has great IDE support.

**Comment by samreid** [comment]
The new pattern is working well, and we added better type checking for it in https://github.com/phetsims/tasks/issues/1132, closing.
